File: monitor_app/lib/screens/monitor_screen.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../models/monitor_client.dart';
import '../models/send_state.dart';
import '../services/credentials_store.dart';
import '../services/osc_bridge.dart';
import '../widgets/channel_strip.dart';
import '../widgets/fader_widget.dart';
import 'connection_screen.dart';

class MonitorScreen extends StatefulWidget {
  final OscBridge bridge;
  const MonitorScreen({super.key, required this.bridge});

  @override
  State<MonitorScreen> createState() => _MonitorScreenState();
}

class _MonitorScreenState extends State<MonitorScreen> {
  late StreamSubscription<BridgeEvent> _sub;
  int _tabIndex = 0; // 0 = My Mix, 1 = My Aux

  @override
  void initState() {
    super.initState();
    _sub = widget.bridge.events.listen(_onEvent);
    // Pull whatever the background service has cached so the UI repaints
    // immediately on this screen's first build (and on every resume).
    widget.bridge.requestSnapshot();
  }

  @override
  void dispose() {
    _sub.cancel();
    super.dispose();
  }

  void _onEvent(BridgeEvent event) {
    if (!mounted) return;
    final model = context.read<MonitorClientModel>();

    switch (event) {
      case StatusChanged(:final connected, :final console):
        if (console.isNotEmpty) {
          model.consoleName = console;
        }
        model.status = connected
            ? ConnectionStatus.connected
            : ConnectionStatus.connecting;
        model.notifyListeners();

      case Snapshot s:
        if (s.console.isNotEmpty) model.consoleName = s.console;
        model.status = s.connected
            ? ConnectionStatus.connected
            : ConnectionStatus.connecting;
        // Replay names first so subsequent send entries pick them up.
        for (final n in s.names) {
          _applyName(model, n);
        }
        for (final send in s.sends) {
          model.updateSend(send.input, send.aux, send.level, send.pan, send.on);
          if (!model.permittedAuxes.contains(send.aux)) {
            model.permittedAuxes.add(send.aux);
          }
        }
        for (final aux in s.auxes) {
          model.updateAux(aux.aux, aux.fader, aux.mute);
        }
        model.permittedAuxes.sort();
        model.selectedAux ??=
            model.permittedAuxes.isNotEmpty ? model.permittedAuxes.first : null;
        model.notifyListeners();

      case SendUpdated(
          :final input,
          :final aux,
          :final level,
          :final pan,
          :final on,
        ):
        model.updateSend(input, aux, level, pan, on);
        if (!model.permittedAuxes.contains(aux)) {
          model.permittedAuxes.add(aux);
          model.permittedAuxes.sort();
          model.selectedAux ??= aux;
          model.notifyListeners();
        }

      case AuxUpdated(:final aux, :final fader, :final mute):
        model.updateAux(aux, fader, mute);

      case ChannelNamed n:
        _applyName(model, n);
        model.notifyListeners();
    }
  }

  void _applyName(MonitorClientModel model, ChannelNamed n) {
    if (n.kind == 'input') {
      for (final entry in model.sends.entries) {
        if (entry.key.$1 == n.ch) {
          entry.value.name = n.name;
        }
      }
    } else if (n.kind == 'aux') {
      final state = model.auxStates.putIfAbsent(
        n.ch,
        () => AuxState(auxCh: n.ch),
      );
      state.name = n.name;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Consumer<MonitorClientModel>(
      builder: (context, model, _) {
        return Scaffold(
          backgroundColor: const Color(0xFF0D0D0D),
          appBar: AppBar(
            backgroundColor: const Color(0xFF1A1A1A),
            title: Row(
              children: [
                Text(
                  model.consoleName.isNotEmpty
                      ? model.consoleName
                      : 'S21 Monitor',
                  style: const TextStyle(fontSize: 16),
                ),
                const SizedBox(width: 12),
                Container(
                  padding:
                      const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
                  decoration: BoxDecoration(
                    color: model.status == ConnectionStatus.connected
                        ? Colors.green
                        : Colors.red,
                    borderRadius: BorderRadius.circular(4),
                  ),
                  child: Text(
                    model.clientName,
                    style: const TextStyle(fontSize: 12, color: Colors.white),
                  ),
                ),
              ],
            ),
            actions: [
              IconButton(
                icon: const Icon(Icons.power_settings_new, size: 20),
                tooltip: 'Shut down service',
                onPressed: () => _shutdown(model),
              ),
            ],
          ),
          body: Column(
            children: [
              Container(
                color: const Color(0xFF1A1A1A),
                child: Row(
                  children: [
                    _tabButton('My Mix', 0),
                    _tabButton('My Aux', 1),
                    const Spacer(),
                    if (_tabIndex == 0 && model.availableAuxes.isNotEmpty)
                      Padding(
                        padding: const EdgeInsets.symmetric(horizontal: 8),
                        child: DropdownButton<int>(
                          value: model.selectedAux,
                          hint: const Text('Select Aux',
                              style: TextStyle(color: Colors.white54)),
                          dropdownColor: const Color(0xFF1A1A1A),
                          style: const TextStyle(color: Colors.white),
                          items: model.availableAuxes.map((a) {
                            final auxState = model.auxStates[a];
                            final auxName = auxState?.name ?? '';
                            final label = auxName.isNotEmpty
                                ? '$auxName (Aux $a)'
                                : 'Aux $a';
                            return DropdownMenuItem(
                              value: a,
                              child: Text(label),
                            );
                          }).toList(),
                          onChanged: (v) {
                            setState(() => model.selectedAux = v);
                          },
                        ),
                      ),
                  ],
                ),
              ),
              Expanded(
                child: _tabIndex == 0
                    ? _buildMyMix(context, model)
                    : _buildMyAux(context, model),
              ),
            ],
          ),
        );
      },
    );
  }

  Widget _tabButton(String label, int index) {
    final selected = _tabIndex == index;
    return GestureDetector(
      onTap: () => setState(() => _tabIndex = index),
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 12),
        decoration: BoxDecoration(
          border: Border(
            bottom: BorderSide(
              color: selected ? Colors.blueAccent : Colors.transparent,
              width: 2,
            ),
          ),
        ),
        child: Text(
          label,
          style: TextStyle(
            color: selected ? Colors.white : Colors.white54,
            fontWeight: selected ? FontWeight.bold : FontWeight.normal,
          ),
        ),
      ),
    );
  }

  // ── My Mix tab ──

  Widget _buildMyMix(BuildContext context, MonitorClientModel model) {
    if (model.selectedAux == null) {
      return const Center(
        child: Text(
          'Select an aux to start mixing',
          style: TextStyle(color: Colors.white54),
        ),
      );
    }

    final sends = model.sendsForSelectedAux();
    if (sends.isEmpty) {
      return const Center(
        child: Text(
          'Waiting for state from daemon...',
          style: TextStyle(color: Colors.white54),
        ),
      );
    }

    final isWide = MediaQuery.of(context).size.width >= 600;

    if (isWide) {
      return SingleChildScrollView(
        scrollDirection: Axis.horizontal,
        padding:
            const EdgeInsets.only(left: 8, right: 8, top: 8, bottom: 40),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: sends.map((send) {
            return Padding(
              padding: const EdgeInsets.symmetric(horizontal: 2),
              child: SizedBox(
                height: double.infinity,
                child: VerticalChannelStrip(
                  send: send,
                  onLevelChanged: (v) => _sendLevel(model, send, v),
                  onPanChanged: (v) => _sendPan(model, send, v),
                  onToggle: () => _sendToggle(model, send),
                ),
              ),
            );
          }).toList(),
        ),
      );
    } else {
      return ListView.separated(
        padding: const EdgeInsets.all(8),
        itemCount: sends.length,
        separatorBuilder: (_, __) => const SizedBox(height: 4),
        itemBuilder: (_, i) {
          final send = sends[i];
          return HorizontalChannelStrip(
            send: send,
            onLevelChanged: (v) => _sendLevel(model, send, v),
            onPanChanged: (v) => _sendPan(model, send, v),
            onToggle: () => _sendToggle(model, send),
          );
        },
      );
    }
  }

  // ── My Aux tab ──

  Widget _buildMyAux(BuildContext context, MonitorClientModel model) {
    final auxes = model.availableAuxes;
    if (auxes.isEmpty) {
      return const Center(
        child: Text(
          'No aux channels assigned',
          style: TextStyle(color: Colors.white54),
        ),
      );
    }

    return ListView(
      padding: const EdgeInsets.all(16),
      children: auxes.map((auxCh) {
        final auxState = model.auxStates[auxCh];
        final fader = auxState?.fader ?? -150.0;
        final mute = auxState?.mute ?? false;
        final auxName = auxState?.name ?? '';
        final displayName =
            auxName.isNotEmpty ? '$auxName (Aux $auxCh)' : 'Aux $auxCh';

        return Container(
          margin: const EdgeInsets.only(bottom: 12),
          padding: const EdgeInsets.all(16),
          decoration: BoxDecoration(
            color: const Color(0xFF1A1A1A),
            borderRadius: BorderRadius.circular(8),
          ),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                displayName,
                style: TextStyle(
                  color: mute ? Colors.white30 : Colors.white,
                  fontSize: 16,
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 8),
              Row(
                children: [
                  Expanded(
                    child: HorizontalFader(
                      value: fader,
                      dbMin: -80.0,
                      dbMax: 10.0,
                      active: !mute,
                      onChanged: (v) {
                        model.updateAux(auxCh, v, mute);
                        widget.bridge.setAuxFader(auxCh, v);
                      },
                    ),
                  ),
                  const SizedBox(width: 8),
                  Text(
                    _formatDb(fader),
                    style: TextStyle(
                        color: mute ? Colors.white24 : Colors.white54),
                  ),
                  const SizedBox(width: 12),
                  GestureDetector(
                    onTap: () {
                      final newMute = !mute;
                      model.updateAux(auxCh, fader, newMute);
                      widget.bridge.setAuxMute(auxCh, newMute);
                    },
                    child: Container(
                      width: 56,
                      height: 32,
                      decoration: BoxDecoration(
                        color: mute ? Colors.red : Colors.grey[800],
                        borderRadius: BorderRadius.circular(4),
                      ),
                      child: Center(
                        child: Text(
                          'MUTE',
                          style: TextStyle(
                            color: mute ? Colors.white : Colors.white54,
                            fontWeight: FontWeight.bold,
                            fontSize: 11,
                          ),
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ],
          ),
        );
      }).toList(),
    );
  }

  // ── Send helpers ──

  void _sendLevel(MonitorClientModel model, SendState send, double value) {
    send.level = value;
    model.notifyListeners();
    widget.bridge.setSendLevel(send.inputCh, send.auxCh, value);
  }

  void _sendPan(MonitorClientModel model, SendState send, double value) {
    send.pan = value;
    model.notifyListeners();
    widget.bridge.setSendPan(send.inputCh, send.auxCh, value);
  }

  void _sendToggle(MonitorClientModel model, SendState send) {
    send.on = !send.on;
    model.notifyListeners();
    widget.bridge.setSendOn(send.inputCh, send.auxCh, send.on);
  }

  /// Tear down the background service AND clear stored credentials.
  /// Matches the WFS DIY pattern: an explicit "I'm done" action that
  /// stops the OSC daemon connection and returns to the connection
  /// wizard. Subsequent app launches will require entering creds again.
  Future<void> _shutdown(MonitorClientModel model) async {
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (ctx) => AlertDialog(
        title: const Text('Shut down service?'),
        content: const Text(
          'This stops the background OSC connection. The app will return '
          'to the connection screen.',
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.of(ctx).pop(false),
            child: const Text('Cancel'),
          ),
          FilledButton(
            onPressed: () => Navigator.of(ctx).pop(true),
            child: const Text('Shut down'),
          ),
        ],
      ),
    );
    if (confirmed != true) return;

    widget.bridge.shutdown();
    await CredentialsStore.clear();

    if (!mounted) return;
    model.setDisconnected();
    model.sends.clear();
    model.auxStates.clear();
    model.selectedAux = null;
    model.permittedAuxes.clear();

    Navigator.of(context).pushReplacement(
      MaterialPageRoute(
        builder: (_) => ConnectionScreen(bridge: widget.bridge),
      ),
    );
    debugPrint('S21 Monitor: background service shut down');
  }

  String _formatDb(double db) {
    if (db <= -59) return '-inf';
    return '${db.toStringAsFixed(1)} dB';
  }
}

File: monitor_android/app/src/main/java/com/pob31/s21monitor/ui/MonitorScreen.kt
package com.pob31.s21monitor.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.PowerSettingsNew
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.pob31.s21monitor.model.MonitorUiState
import com.pob31.s21monitor.service.MonitorService
import com.pob31.s21monitor.ui.theme.Accent
import com.pob31.s21monitor.ui.theme.Bg
import com.pob31.s21monitor.ui.theme.Danger
import com.pob31.s21monitor.ui.theme.Muted
import com.pob31.s21monitor.ui.theme.OnAccent
import com.pob31.s21monitor.ui.theme.Panel
import com.pob31.s21monitor.ui.theme.Panel2
import com.pob31.s21monitor.ui.theme.TextPrimary
import com.pob31.s21monitor.ui.widgets.AuxStrip
import com.pob31.s21monitor.ui.widgets.SendStrip

@Composable
fun MonitorScreen(
    state: MonitorUiState,
    clientName: String,
    service: MonitorService,
    onShutdown: () -> Unit,
) {
    var tab by remember { mutableIntStateOf(0) } // 0 = My Mix, 1 = My Aux
    var selectedAux by remember { mutableStateOf<Int?>(null) }

    val auxes = state.availableAuxes
    LaunchedEffect(auxes) {
        if (selectedAux == null || selectedAux !in auxes) selectedAux = auxes.firstOrNull()
    }

    Column(Modifier.fillMaxSize().background(Bg)) {
        Header(
            console = state.console.ifEmpty { "S21 Monitor" },
            clientName = clientName,
            connected = state.connected,
            onShutdown = onShutdown,
        )
        TabsRow(
            tab = tab,
            onTab = { tab = it },
            showAuxPicker = tab == 0 && auxes.isNotEmpty(),
            auxes = auxes,
            auxStates = state,
            selectedAux = selectedAux,
            onSelectAux = { selectedAux = it },
        )
        Box(Modifier.fillMaxSize()) {
            if (tab == 0) MyMix(state, selectedAux, service) else MyAux(state, service)
        }
    }
}

@Composable
private fun Header(
    console: String,
    clientName: String,
    connected: Boolean,
    onShutdown: () -> Unit,
) {
    Row(
        Modifier.fillMaxWidth().background(Panel).padding(horizontal = 14.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Text(console, color = TextPrimary, fontSize = 17.sp, fontWeight = FontWeight.SemiBold)
        Text(clientName, color = Muted, fontSize = 13.sp)
        Box(Modifier.weight(1f))
        Box(
            Modifier.size(11.dp).clip(CircleShape)
                .background(if (connected) Accent else Danger),
        )
        Text(
            if (connected) "Connected" else "Connecting…",
            color = Muted, fontSize = 12.sp,
        )
        Icon(
            Icons.Filled.PowerSettingsNew,
            contentDescription = "Shut down",
            tint = Muted,
            modifier = Modifier.size(22.dp).clip(CircleShape).clickable(onClick = onShutdown),
        )
    }
}

@Composable
private fun TabsRow(
    tab: Int,
    onTab: (Int) -> Unit,
    showAuxPicker: Boolean,
    auxes: List<Int>,
    auxStates: MonitorUiState,
    selectedAux: Int?,
    onSelectAux: (Int) -> Unit,
) {
    Row(
        Modifier.fillMaxWidth().background(Bg).padding(horizontal = 12.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Pill("My Mix", tab == 0) { onTab(0) }
        Pill("My Aux", tab == 1) { onTab(1) }
        Box(Modifier.weight(1f))
        if (showAuxPicker) {
            AuxPicker(auxes, auxStates, selectedAux, onSelectAux)
        }
    }
}

@Composable
private fun Pill(label: String, active: Boolean, onClick: () -> Unit) {
    Box(
        Modifier
            .clip(RoundedCornerShape(20.dp))
            .background(if (active) Accent else Panel2)
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        Text(
            label,
            color = if (active) OnAccent else Muted,
            fontSize = 13.sp,
            fontWeight = if (active) FontWeight.SemiBold else FontWeight.Normal,
        )
    }
}

@Composable
private fun AuxPicker(
    auxes: List<Int>,
    state: MonitorUiState,
    selectedAux: Int?,
    onSelectAux: (Int) -> Unit,
) {
    var open by remember { mutableStateOf(false) }
    fun label(aux: Int?): String {
        if (aux == null) return "Select Aux"
        val name = state.auxes[aux]?.name.orEmpty()
        return if (name.isNotEmpty()) "$name (Aux $aux)" else "Aux $aux"
    }
    Box {
        Row(
            Modifier.clip(RoundedCornerShape(8.dp)).background(Panel2)
                .clickable { open = true }.padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(label(selectedAux), color = TextPrimary, fontSize = 13.sp)
            Icon(Icons.Filled.ExpandMore, contentDescription = null, tint = Muted)
        }
        DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            auxes.forEach { a ->
                DropdownMenuItem(text = { Text(label(a)) }, onClick = {
                    onSelectAux(a); open = false
                })
            }
        }
    }
}

@Composable
private fun MyMix(state: MonitorUiState, selectedAux: Int?, service: MonitorService) {
    if (selectedAux == null) {
        Empty("Select an aux to start mixing"); return
    }
    val sends = state.sendsForAux(selectedAux)
    if (sends.isEmpty()) {
        Empty("Waiting for state from the daemon…"); return
    }
    Row(
        Modifier.fillMaxSize().horizontalScroll(rememberScrollState()).padding(10.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        sends.forEach { send ->
            SendStrip(
                send = send,
                onLevel = { service.setSendLevel(send.input, send.aux, it) },
                onPan = { service.setSendPan(send.input, send.aux, it) },
                onToggle = { service.setSendOn(send.input, send.aux, !send.on) },
            )
        }
    }
}

@Composable
private fun MyAux(state: MonitorUiState, service: MonitorService) {
    val auxes = state.availableAuxes
    if (auxes.isEmpty()) {
        Empty("No aux channels assigned"); return
    }
    Row(
        Modifier.fillMaxSize().horizontalScroll(rememberScrollState()).padding(10.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        auxes.forEach { a ->
            val aux = state.auxes[a] ?: com.pob31.s21monitor.model.AuxState(a)
            AuxStrip(
                aux = aux,
                onFader = { service.setAuxFader(a, it) },
                onMute = { service.setAuxMute(a, !aux.mute) },
            )
        }
    }
}

@Composable
private fun Empty(text: String) {
    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        Text(text, color = Muted, fontSize = 14.sp)
    }
}
